workspaces/arborist/test/isolated-mode.js

```diff
@@ -552,9 +552,10 @@ tap.test('failing optional deps are not installed', async t => {
   const arborist = new Arborist({ path: dir, registry, packumentCache: new Map(), cache })
   await arborist.reify({ installStrategy: 'linked' })
 
-  t.notOk(setupRequire(dir)('which'), 'Failing optional deps should not be installed')
+  t.ok(setupRequire(dir)('which'), 'Failing optional deps should not be installed, but the empty directory is there')
```

Comment:
Am I right to say that the new behavior would be to leave empty directories for all of the optional deps that did not succeed? Hopefully there's a way to avoid this, since that would leave a load of empty dirs for a lot of native packages, and potentially confuse code which is scanning for these dirs to know which one to load for the current platform...

Reply:
That is correct indeed. This is used to rebuild the proper package-json.lock file based on the `node_modules` directory.
I don't know enough from NPM to surface a better solution, but maybe others have ideas on how this can prevented?

Reply:
That's really unfortunate; I thought that npm stored a file inside `node_modules` that described the layout? It seems like a bad idea to rely on an empty dir existing, and I really do think it'll cause something to break.

Reply:
Yes, i can image breakage if packages rely on directory scanning.

If this is not the desired approach, i would need guidance from the maintainers on what would be the best way forward.

Reply:
Yep, we'll have @wraithgar weigh in

Reply:
For what it's worth, i checked out `bindings` which i use myself in https://github.com/blueconic/node-oom-heapdump, and that library handles this just fine. Haven't checked how nodejs itself handles it, as it's able to load native modules out of the box nowadays. 

https://github.com/TooTallNate/node-bindings/blob/c8033dcfc04c34397384e23f7399a30e6c13830d/bindings.js#L111

Reply:
The main ones to check out are rollup and esbuild (the bulk of the people in the issue having this problem), but there's also swc, oxc, dprint, who have this same problem too.

Reply:
https://github.com/evanw/esbuild/blob/b914dd30294346aa15fcc04278f4b4b51b8b43b5/lib/npm/node-platform.ts#L150 is fine for the happy path, but the fallback does use directory scanning.